Introduce peep animation (costume) objects

AaronVanGeffen opened the pull request:
This PR introduces peep animation (costume) objects to OpenRCT2. In essence, all sprite offsets for all peeps are moved from `PeepAnimationData.cpp` to objects. This includes all guest animations, as well as animations for all staff types. In the short run, this will allow players to make their own custom entertainer costumes.

To facilitate this, the Peep struct is extended with an `AnimationObjectIndex` property. The existing `AnimationGroup` is reworked to offset on a peep animation object instead. To populate both properties, an import step is added to the save file loading process. This automatically appends the relevant peep animation objects to existing saves, taking care to add entertainers based on what (vanilla) scenery group objects are selected in the save game. Obviously, this needs to be thoroughly tested.

**Note for the curious/adventurous/testers**
This PR depends on https://github.com/OpenRCT2/objects/pull/357, which has a few to-dos left. Until we've made a new object release including the objects, you will need to add these objects to your test setup manually.

**To do**
- [ ] 'Current costume' in staff window isn't showing until dropdown is opened
- [ ] Retrieving available costumes using scripts still uses hardcoded costume lists
- [ ] ...
- [ ] Testing, loads of testing

ZeeMaji commented:
Found a niche bug, if the only selected entertainer costume is one that has noRandomPlacement then you can't hire entertainers at all with this out of range error.
![Screenshot at 2024-12-11 07-49-35](https://github.com/user-attachments/assets/50604ba1-dff1-4572-93c1-e21f2613866d)

AaronVanGeffen commented:
> Found a niche bug, if the only selected entertainer costume is one that has noRandomPlacement then you can't hire entertainers at all with this out of range error.

I'm glad to see it returning a nice error if there are no costumes available, haha. What would you prefer it to do in this case?

ZeeMaji commented:
> I'm glad to see it returning a nice error if there are no costumes available, haha. What would you prefer it to do in this case?

It would probably make sense to just ignore the flag if the only available entertainers all have the flag enabled. It's a pretty niche situation since it requires you to deliberately disable all random available entertainer objects. Or some other similar solution, i think it'd be fine as long as it lets you actually hire entertainers in this niche situation.

AaronVanGeffen commented:
@ZeeMaji Filed a patch to address the issue.

ZeeMaji commented:
The window tab walking animation still seems to be off, it goes too fast now and it still doesn't seem to properly handle long animation sequences.

https://github.com/user-attachments/assets/487b1384-1f45-43a6-9c5a-a745e963dcfd

AaronVanGeffen commented:
> The window tab walking animation still seems to be off, it goes too fast now and it still doesn't seem to properly handle long animation sequences.

Is it better now, as of [31e6c48](https://github.com/OpenRCT2/OpenRCT2/pull/23328/commits/31e6c4864e19573124c4ff495440a31070fcae94)?

AaronVanGeffen commented:
Ahh, that's right! Thanks for noticing that. The same thing will need doing for the pickup sequence, I just realised. I'll make both use their respective frame sequences.

ZeeMaji commented:
It would probably make sense to add the elephant, panda, and tiger costumes as pre-selected objects when creating a new scenario from the main menu toolbox menu as presently you get an error that no entertainers are selected if you try to proceed to the next editor step without selecting any.

ZeeMaji commented:
It would probably make sense to add the elephant, panda, and tiger costumes as pre-selected objects when creating a new scenario from the main menu toolbox menu as presently you get an error that no entertainers are selected if you try to proceed to the next editor step without selecting any.

AaronVanGeffen commented:
They're intended to be inherited from the trees scenery group, as was the case before. Did you test with outdated scenery group objects, by any chance?

ZeeMaji commented:
Ah right, that'd explain it. I only added the peep_animation objects as i'm testing this via the packed appimage which doesn't have all of the object updates yet.

ZeeMaji commented:
An issue i noticed, but security guard behavior seems to have changed. When they change to the "hands behind back" walking animation they don't slow down like they're supposed to.

AaronVanGeffen commented:
> An issue i noticed, but security guard behavior seems to have changed. When they change to the "hands behind back" walking animation they don't slow down like they're supposed to.

Interesting. I don't think I've touched anything like that for security guards. I did notice this when I was reimplementing entertainer costume assignment, though, hence the slow walking property.

Is this working correctly in develop, or are you comparing this to vanilla?

ZeeMaji commented:
I was comparing to develop, after checking vanilla it exists there too.

AaronVanGeffen commented:
This bit needs to be moved to the relevant animations in the objects concerned:
https://github.com/OpenRCT2/OpenRCT2/blob/7baae607ac57eb3dd2132d8a105852f4213c1605/src/openrct2/entity/Peep.cpp#L123-L127

After which this bit will need to be adjusted to use the object:
https://github.com/OpenRCT2/OpenRCT2/blob/7baae607ac57eb3dd2132d8a105852f4213c1605/src/openrct2/entity/Staff.cpp#L1713-L1717

As well as this bit:
https://github.com/OpenRCT2/OpenRCT2/blob/7baae607ac57eb3dd2132d8a105852f4213c1605/src/openrct2/entity/Guest.cpp#L6768-L6773

The `gAnimationGroupToSlowWalkMap` array can then be removed.

AaronVanGeffen commented:
With the objects and replay bundles updated, I believe this is finally ready for review.

ZeeMaji commented:
Sitting guests with umbrellas have the umbrellas too high, chopping half of their head off.
![Screenshot at 2024-12-25 05-37-20](https://github.com/user-attachments/assets/9d4710f1-1aec-4413-9593-5b3747f9db24)

AaronVanGeffen commented:
New replay recordings are required to account for #23200 having been merged...

AaronVanGeffen commented:
@ZeeMaji The issue with umbrellas/hats/balloons for sitting peeps should now be fixed.

I've also re-recorded the replays to account for the park version also having been bumped by https://github.com/OpenRCT2/OpenRCT2/pull/23200.

ZeeMaji commented:
A minor issue i noticed, but the object selection window doesn't render remappable sprites properly. This applies to the guest, handyman, mechanic, and security guard previews. Primary remap renders like chain (as the special orange values it uses doesn't exist in the normal palette) while the secondary remap (just guest pants presently) are pink.
![Screenshot at 2024-12-27 08-42-17](https://github.com/user-attachments/assets/f11e5d04-ea74-49af-96c9-522ddbb936a1)

Not sure on what the best colors for the preview would be especially since we'll probably support more peep types with secondary remaps and possibly even tertiary remaps at some point. I think the safest bet would be to just make the primary remap render as the normal orange and leave the secondary as is (pink).

ZeeMaji commented:
Also i'm not sure if it's specific to this branch, but i've noticed that you can bypass the minimum object requirements that are set by using the backspace button to force close the object selection window which is pretty bad here since you can disable guests or non-entertainer staff types which will result in crashing.

AaronVanGeffen commented:
> Not sure on what the best colors for the preview would be especially since we'll probably support more peep types with secondary remaps and possibly even tertiary remaps at some point. I think the safest bet would be to just make the primary remap render as the normal orange and leave the secondary as is (pink).

I plan on implementing recolourable entertainers in the near future, which would allow setting the default remap colours in the peep animation object. Until then, I think we'll have to implement a global default at least, yes.

> Also i'm not sure if it's specific to this branch, but i've noticed that you can bypass the minimum object requirements that are set by using the backspace button to force close the object selection window which is pretty bad here since you can disable guests or non-entertainer staff types which will result in crashing.

That's a nasty oversight that affects develop as well. I'll file a patch.

ZeeMaji commented:
A UX improvement i'd like to see is to display what peep type the currently viewed object is above the source game in the bottom right of the object selection window ala ride types for ride vehicle objects.
Note the little "3D Cinema" text above the source game.
![Screenshot at 2024-12-29 07-22-28](https://github.com/user-attachments/assets/5d0951b6-5c16-4ecc-a79f-1a72cd0cc288)

I think it would be nice if it would display "Guest", "Entertainer", "Handyman", etc... above the source game. That way it'll be easier to identify what type of peeps objects are since for custom objects it might not be obvious.
![Screenshot at 2024-12-29 07-22-38](https://github.com/user-attachments/assets/9958bc09-aef4-4ccb-a9bb-6c17afea0b31)

AaronVanGeffen commented:
@AaronVanGeffen Added:
![Unnamed park 2024-12-29 13-47-49](https://github.com/user-attachments/assets/4f0fcb73-32c1-4d8b-8ed1-a4f0700f55ed)

I'd really like to get this in before the next release. Unless there's anything game-breaking, please keep additional requests for after this PR.

ZeeMaji commented:
When importing old parks that have entertainers from unavailable scenery groups (eg: roman entertainer with no roman theming) they will not get converted properly and will get converted as the lowest index entertainer (usually the elephant costume)
This probably also affects s4 and s6 but i do not have any saves that have this setup that i'm aware of.
Forest frontiers save with all of the entertainers used without their scenery groups enabled:
[Forest Frontiersroman3.park.tar.gz](https://github.com/user-attachments/files/18269844/Forest.Frontiersroman3.park.tar.gz)

ZeeMaji commented:
The walking guest tab icons in the guest windows doesn't display properly for guests with hats/balloons.
![Screenshot at 2024-12-30 17-23-30](https://github.com/user-attachments/assets/2495a2e5-314e-4c80-968f-be194fae56db)
![Screenshot at 2024-12-30 17-20-08](https://github.com/user-attachments/assets/660e52e1-ab3f-4021-af16-d89f4688c3e8)

Gymnasiast commented on src/openrct2-ui/windows/Staff.cpp:
```diff
@@ -768,31 +793,28 @@ namespace OpenRCT2::Ui::Windows
+            auto numCostumes = _availableCostumes.size();
+            for (auto i = 0u; i < numCostumes; i++)
+            {
+                // TODO: refactor this hack
```
How and when?

AaronVanGeffen replied:
Currently, dropdown args are defined as an `int64_t`. This means the `char*` is assigned using `memcpy`, so it works with both 32-bit and 64-bit builds. It feels incredibly hacky to me, but it is how it works in several places in the codebase.

Ideally, there'd be a more universal (variant) interface for dropdown arguments. I'll amend the comment to clarify.

Gymnasiast commented on src/openrct2-ui/windows/StaffList.cpp:
```diff
@@ -549,7 +549,15 @@ namespace OpenRCT2::Ui::Windows
             }
 
-            auto hireStaffAction = StaffHireNewAction(autoPosition, staffType, entertainerType, staffOrders);
+            // Select a (random) costume for this staff member
```
Already clear from the function names.

Gymnasiast commented on src/openrct2/EditorObjectSelectionSession.cpp:
```diff
@@ -248,6 +250,25 @@ void SetupInUseSelectionFlags()
         Editor::SetSelectedObject(ObjectType::Music, ride.music, ObjectSelectionFlags::InUse);
     }
 
+    // Figure out what peep animations are in use
```
Already clear from the code.

Gymnasiast commented on src/openrct2/interface/InteractiveConsole.cpp:
```diff
@@ -1163,7 +1166,8 @@ constexpr std::array _objectTypeNames = {
     "Audio",
-    "Guest Names",
+    "Peep Names",
+    "Peep Animations",
```
It doesn’t need to be changed for this PR, but this should just use the translated names that the Object Selection uses. (If we even keep the console.)

AaronVanGeffen replied:
Indeed, the whole thing is pretty much not translated right now. This could be changed to map to string ids with relative ease, though. Perhaps this could be a 'good first issue'?

Gymnasiast commented on src/openrct2/object/PeepAnimationsObject.cpp:
```diff
@@ -0,0 +1,226 @@
+
+        PeepAnimations group = ReadAnimations(requiredAnimationMap, groupJson["animations"]);
+
+        // Is this animation group replacing a legacy group?
```
Again, already clear from the code itself.

Comments should **add** something, rather than stating the obvious.

AaronVanGeffen replied:
Haha, I think we just have a different code style here. Personally, I like to have narrative comments in my code, as I think it makes it easier to read for newcomers. As such, I think they do add something.

I checked with some people on Discord, and it seems it's clearly not common for this codebase. I'll remove the 'obvious' comments you've pointed out.

Gymnasiast commented on src/openrct2/object/PeepAnimationsObject.cpp:
```diff
@@ -0,0 +1,226 @@
+            group.legacyPosition = RCT12PeepAnimationGroup::Invalid;
+        }
+
+        // Should we play back the walking animation more slowly?
```
Idem

Gymnasiast commented on src/openrct2/object/PeepAnimationsObject.cpp:
```diff
@@ -0,0 +1,226 @@
+        // Should we play back the walking animation more slowly?
+        group.isSlowWalking = Json::GetBoolean(groupJson["isSlowWalking"], false);
+
+        // Do we have a preferred way of addressing this object in scripts?
```
Idem

Gymnasiast commented on src/openrct2/park/ParkFile.cpp:
```diff
@@ -385,6 +385,13 @@ namespace OpenRCT2
 
+                if (version < kPeepAnimationObjectsVersion)
+                {
+                    // Add all legacy peep animation objects
```
Idem

Gymnasiast commented on src/openrct2/rct2/S6Importer.cpp:
```diff
@@ -1900,6 +1905,10 @@ namespace OpenRCT2::RCT2
                 AppendRequiredObjects(
                     objectList, ObjectType::ScenarioText, std::vector<std::string_view>({ desc.textObjectId }));
 
+            // Add all legacy peep animation objects
```
Idem

Gymnasiast commented on src/openrct2/scripting/bindings/entity/ScStaff.cpp:
```diff
@@ -168,29 +137,46 @@ namespace OpenRCT2::Scripting
-        auto& availableCostumes = costumesByStaffType(peep->AssignedStaffType);
+        auto& costumes = costumesByStaffType(peep->AssignedStaffType);
+
+        // Find the peep's current costume in the list
```
Idem

Gymnasiast commented on src/openrct2/scripting/bindings/entity/ScStaff.cpp:
```diff
@@ -203,30 +189,30 @@ namespace OpenRCT2::Scripting
-                peep->AnimationGroup = *newAnimationGroup;
-                return;
-            }
+            // Find the peep's current costume by string
```
Idem

Gymnasiast commented on src/openrct2/scripting/bindings/entity/ScStaff.cpp:
```diff
@@ -203,30 +189,30 @@ namespace OpenRCT2::Scripting
-                peep->AnimationGroup = newAnimationGroup;
-                return;
-            }
+            // Find the peep's current costume by legacy number
```
Idem

Gymnasiast commented:
Let’s get this bitch merged. I looked through the code and apart from dozens of comments that IMO just state the obvious, as well as one TODO, I didn’t see anything particularly nasty. If you could address the comment, rebase the PR I’ll take it for a spin.

AaronVanGeffen commented:
Thanks for the review. I believe I have addressed the comments. A changelog entry still needs to be added, but I'll wait for a bit so the branch doesn't conflict again.

Note to self: before merging, we should re-tag the object bundle release to include the latest set of translations.

Gymnasiast commented:
I did a quick test.

I have one thing: this changes the icon of the Entertainers in the Staff window to an astronaut, presumably because it’s the first available type. Would like to see the order changed so this defaults to a panda (RCT1) or an elephant (RCT2, OpenRCT2).

Apart from that, it looks good, though I would like to see this tested by the DKMP crowd.

AaronVanGeffen commented:
> I have one thing: this changes the icon of the Entertainers in the Staff window to an astronaut, presumably because it’s the first available type. Would like to see the order changed so this defaults to a panda (RCT1) or an elephant (RCT2, OpenRCT2).

AaronVanGeffen commented:
> I have one thing: this changes the icon of the Entertainers in the Staff window to an astronaut, presumably because it’s the first available type. Would like to see the order changed so this defaults to a panda (RCT1) or an elephant (RCT2, OpenRCT2).

Hmm, this was a deliberate change, though. The entertainers you mentioned can be omitted, after all. For legacy parks, I can make it such that the objects for panda/elephant/tiger entertainers are appended before the other entertainers. I think that would alleviate your concerns, right?

(For parks that are newly created through the scenario editor, only the panda/elephant/tiger entertainers are available by default, so that should be fine already.)

> Apart from that, it looks good, though I would like to see this tested by the DKMP crowd.

Sure. Could you invite them, then?

Gymnasiast commented:
> Hmm, this was a deliberate change, though. The entertainers you mentioned can be omitted, after all. For legacy parks, I can make it such that the objects for panda/elephant/tiger entertainers are appended before the other entertainers. I think that would alleviate your concerns, right?

Certainly - I was just talking about existing parks. I think that is just a modification of one or two tables, correct?

AaronVanGeffen commented:
> Certainly - I was just talking about existing parks. I think that is just a modification of one or two tables, correct?

Yes. That should be addressed by [b83e99c](https://github.com/OpenRCT2/OpenRCT2/pull/23328/commits/b83e99c81003a86464cd4e0b75f93591bb29e186) then.

deurklink commented:
Aaron asked me to this this PR. I checked the guests in several of my parks and found nothing out of the ordinary. I also used the Peep Editor plugin beta to cycle through all the animations of a guest from all 4 angles, and they all seem to work fine. Did the same for a handyman, mechanic, security guard, and several of the entertainer costumes. Didn't see any animations that look weird or broken.

Manticore-007 commented:
I agree with Deurklink, I also gave the Peep Editor a spin, and didn't find anything that is broken. Looks good to me. I was a bit confused at first that the `no costume` option for entertainers was vanished. But Aaron reminded me that we agreed to remove this since it is now possible to edit guests. So the `no costume` option is obsolete.

J-PIE-314 commented:
The word "_costumes_" was mistyped when appending the changelog (26ea60131772fbe4dc2f218b8947d5a487ef2e70).

`Feature: [#23328] Introducing peep animation objects, enabling custom entertainer customes.`
